**Adım 1: Yükleme ve ilk bakış**

In [1]:
import numpy as np
import pandas as pd

# Tüm sütunları görebilmek için gösterim ayarları
pd.set_option("display.max_columns", 100)
pd.set_option("display.max_rows", 100)

TRANSACTION_PATH = "../data/train_transaction.csv"
IDENTITY_PATH = "../data/train_identity.csv"

df_trans = pd.read_csv(TRANSACTION_PATH)
df_id = pd.read_csv(IDENTITY_PATH)


In [2]:
# Boyutlar
print("Transaction shape:", df_trans.shape)
print("Identity shape:   ", df_id.shape)

Transaction shape: (590540, 394)
Identity shape:    (144233, 41)


In [3]:
# Bellek kullanımı (MB)
mem_trans = df_trans.memory_usage(deep=True).sum() / 1024**2
mem_id = df_id.memory_usage(deep=True).sum() / 1024**2
print(f"Transaction memory: {mem_trans:,.1f} MB")
print(f"Identity memory:    {mem_id:,.1f} MB")

Transaction memory: 1,791.7 MB
Identity memory:    56.5 MB


In [4]:
df_trans.head(3)

,TransactionID,isFraud,TransactionDT,TransactionAmt,ProductCD,card1,card2,card3,card4,card5,card6,addr1,addr2,dist1,dist2,P_emaildomain,R_emaildomain,C1,C2,C3,C4,C5,C6,C7,C8,C9,C10,C11,C12,C13,C14,D1,D2,D3,D4,D5,D6,D7,D8,D9,D10,D11,D12,D13,D14,D15,M1,M2,M3,M4,...,V290,V291,V292,V293,V294,V295,V296,V297,V298,V299,V300,V301,V302,V303,V304,V305,V306,V307,V308,V309,V310,V311,V312,V313,V314,V315,V316,V317,V318,V319,V320,V321,V322,V323,V324,V325,V326,V327,V328,V329,V330,V331,V332,V333,V334,V335,V336,V337,V338,V339
0,2987000,0,86400,68.5,W,13926,NaN,150.0,discover,142.0,credit,315.0,87.0,19.0,NaN,NaN,NaN,1.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,2.0,0.0,1.0,1.0,14.0,NaN,13.0,NaN,NaN,NaN,NaN,NaN,NaN,13.0,13.0,NaN,NaN,NaN,0.0,T,T,T,M2,...,1.0,1.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,117.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,117.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2987001,0,86401,29.0,W,2755,404.0,150.0,mastercard,102.0,credit,325.0,87.0,NaN,NaN,gmail.com,NaN,1.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,1.0,0.0,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,M0,...,1.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2987002,0,86469,59.0,W,4663,490.0,150.0,visa,166.0,debit,330.0,87.0,287.0,NaN,outlook.com,NaN,1.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,1.0,0.0,1.0,1.0,0.0,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,0.0,315.0,NaN,NaN,NaN,315.0,T,T,T,M0,...,1.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [5]:
df_id.head(3)

,TransactionID,id_01,id_02,id_03,id_04,id_05,id_06,id_07,id_08,id_09,id_10,id_11,id_12,id_13,id_14,id_15,id_16,id_17,id_18,id_19,id_20,id_21,id_22,id_23,id_24,id_25,id_26,id_27,id_28,id_29,id_30,id_31,id_32,id_33,id_34,id_35,id_36,id_37,id_38,DeviceType,DeviceInfo
0,2987004,0.0,70787.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,100.0,NotFound,NaN,-480.0,New,NotFound,166.0,NaN,542.0,144.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,New,NotFound,Android 7.0,samsung browser 6.2,32.0,2220x1080,match_status:2,T,F,T,T,mobile,SAMSUNG SM-G892A Build/NRD90M
1,2987008,-5.0,98945.0,NaN,NaN,0.0,-5.0,NaN,NaN,NaN,NaN,100.0,NotFound,49.0,-300.0,New,NotFound,166.0,NaN,621.0,500.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,New,NotFound,iOS 11.1.2,mobile safari 11.0,32.0,1334x750,match_status:1,T,F,F,T,mobile,iOS Device
2,2987010,-5.0,191631.0,0.0,0.0,0.0,0.0,NaN,NaN,0.0,0.0,100.0,NotFound,52.0,NaN,Found,Found,121.0,NaN,410.0,142.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Found,Found,NaN,chrome 62.0,NaN,NaN,NaN,F,F,T,T,desktop,Windows


**Adım 2: Sütun yapısını anlama**

In [6]:
# Sütunları ön eklerine göre gruplandırıyoruz
# filter(regex=...) ile "C1, C2, ..., C14" gibi isimleri tek seferde yakalıyoruz
# ^C\d+$ : C ile başlayıp sadece rakamlarla biten isimler (ProductCD gibi sütunları yanlışlıkla yakalamasın diye)
groups = {
    "C (sayaç)":      df_trans.filter(regex=r"^C\d+$").columns,
    "D (zaman farkı)": df_trans.filter(regex=r"^D\d+$").columns,
    "M (eşleşme T/F)": df_trans.filter(regex=r"^M\d+$").columns,
    "V (Vesta özellikleri)": df_trans.filter(regex=r"^V\d+$").columns,
    "card":           df_trans.filter(regex=r"^card\d+$").columns,
}

for name, cols in groups.items():
    print(f"{name:25s}: {len(cols):3d} sütun")

# Gruplara girmeyen, "tek başına" sütunlar
grouped = set().union(*[set(c) for c in groups.values()])
others = [c for c in df_trans.columns if c not in grouped]
print("\nDiğer sütunlar:", others)

C (sayaç)                :  14 sütun
D (zaman farkı)          :  15 sütun
M (eşleşme T/F)          :   9 sütun
V (Vesta özellikleri)    : 339 sütun
card                     :   6 sütun

Diğer sütunlar: ['TransactionID', 'isFraud', 'TransactionDT', 'TransactionAmt', 'ProductCD', 'addr1', 'addr2', 'dist1', 'dist2', 'P_emaildomain', 'R_emaildomain']


In [8]:
# Veri tiplerinin dağılımı: kaç sütun float, int, object (metin)?
print(df_trans.dtypes.value_counts())

# Metin (object) sütunlar kategorik değişkenlerdir ve bellekte en pahalı tip budur
# Hem transaction hem identity için bunları listeliyoruz
obj_trans = df_trans.select_dtypes(include=["object","str"]).columns.tolist()
obj_id = df_id.select_dtypes(include=["object","str"]).columns.tolist()
print("\nTransaction object sütunlar:", obj_trans)
print("Identity object sütunlar:   ", obj_id)

float64    376
str         14
int64        4
Name: count, dtype: int64

Transaction object sütunlar: ['ProductCD', 'card4', 'card6', 'P_emaildomain', 'R_emaildomain', 'M1', 'M2', 'M3', 'M4', 'M5', 'M6', 'M7', 'M8', 'M9']
Identity object sütunlar:    ['id_12', 'id_15', 'id_16', 'id_23', 'id_27', 'id_28', 'id_29', 'id_30', 'id_31', 'id_33', 'id_34', 'id_35', 'id_36', 'id_37', 'id_38', 'DeviceType', 'DeviceInfo']


In [9]:
# Kategorik sütunların kaç farklı değeri var? (cardinality)
# Az değerli sütunlar "category" tipine çevrilince çok yer kazandırır (Adım 5'te kullanacağız)
# Çok değerli olanlar (ör. DeviceInfo) ise ayrı bir strateji gerektirir
card_trans = df_trans[obj_trans].nunique().sort_values(ascending=False)
card_id = df_id[obj_id].nunique().sort_values(ascending=False)
print(card_trans, "\n")
print(card_id)

R_emaildomain    60
P_emaildomain    59
ProductCD         5
card4             4
card6             4
M4                3
M1                2
M2                2
M3                2
M5                2
M6                2
M7                2
M8                2
M9                2
dtype: int64 

DeviceInfo    1786
id_33          260
id_31          130
id_30           75
id_34            4
id_23            3
id_15            3
id_12            2
id_27            2
id_28            2
id_29            2
id_16            2
id_35            2
id_36            2
id_37            2
id_38            2
DeviceType       2
dtype: int64


In [10]:
# Birleştirmeye hazırlık: anahtar sütun (TransactionID) kontrolü
# 1) Transaction'da her ID bir kez mi geçiyor? (unique olmalı)
print("Transaction ID unique mi?:", df_trans["TransactionID"].is_unique)
print("Identity ID unique mi?:   ", df_id["TransactionID"].is_unique)

# 2) Identity'deki ID'lerin hepsi transaction'da var mı?
# isin() ile kontrol ediyoruz; hepsi True ise identity tamamen transaction'ın alt kümesidir
in_trans = df_id["TransactionID"].isin(df_trans["TransactionID"])
print("Identity ID'lerinin transaction'da bulunma oranı:", in_trans.mean())

# 3) Transaction'ların yüzde kaçının kimlik kaydı var?
has_id = df_trans["TransactionID"].isin(df_id["TransactionID"])
print(f"Kimlik kaydı olan işlem oranı: {has_id.mean():.1%}")

Transaction ID unique mi?: True
Identity ID unique mi?:    True
Identity ID'lerinin transaction'da bulunma oranı: 1.0
Kimlik kaydı olan işlem oranı: 24.4%


**Adım 3: Tekrarlı kayıtlar**

In [11]:
# 1) ID hariç tüm sütunlar birebir aynı olan satırlar var mı?
# TransactionID zaten benzersiz olduğu için onu dahil edersek sonuç hep 0 çıkar, o yüzden çıkarıyoruz
# duplicated() varsayılan olarak ilk görüleni "tekrar değil", sonrakileri "tekrar" sayar
cols_wo_id = [c for c in df_trans.columns if c != "TransactionID"]
dup_full = df_trans.duplicated(subset=cols_wo_id)
print("Transaction - ID hariç birebir aynı satır:", dup_full.sum())

# Identity için aynı kontrol
id_cols_wo_id = [c for c in df_id.columns if c != "TransactionID"]
print("Identity - ID hariç birebir aynı satır:   ", df_id.duplicated(subset=id_cols_wo_id).sum())

Transaction - ID hariç birebir aynı satır: 3
Identity - ID hariç birebir aynı satır:    3767


In [12]:
# 3 satırın hangileri olduğunu görelim
# keep=False: tekrar eden grubun TÜM üyelerini işaretler (ilk görüleni de), böylece çiftleri yan yana görürüz
mask = df_trans.duplicated(subset=cols_wo_id, keep=False)

# Aynı gruptaki satırlar alt alta gelsin diye sıralıyoruz
# 3 tekrar var ama grup sayısı 3'ten az olabilir (bir grupta 3 kopya da olabilir)
show_cols = ["TransactionID", "TransactionDT", "TransactionAmt", "card1", "ProductCD", "isFraud"]
df_trans.loc[mask, show_cols].sort_values(["TransactionDT", "card1", "TransactionAmt"])

,TransactionID,TransactionDT,TransactionAmt,card1,ProductCD,isFraud
295456,3282456,7310775,47.95,7919,W,0
295457,3282457,7310775,47.95,7919,W,0
453414,3440414,11576951,512.95,1359,W,0
453415,3440415,11576951,512.95,1359,W,0
570433,3557433,15129302,44.00,2744,W,0
570434,3557434,15129302,44.00,2744,W,0


In [13]:
# Her çiftte ilk kaydı tutup ikincisini atacağız
# keep="first": ilk görüleni "tekrar değil", sonrakini "tekrar" sayar
dup_mask = df_trans.duplicated(subset=cols_wo_id, keep="first")
ids_to_drop = df_trans.loc[dup_mask, "TransactionID"]
print("Silinecek ID'ler:", ids_to_drop.tolist())

# Bu ID'lerin identity tablosunda kaydı var mı? (varsa kimlik bilgisi kaybolur)
print("Identity'de kaydı olan:", ids_to_drop.isin(df_id["TransactionID"]).sum())

Silinecek ID'ler: [3282457, 3440415, 3557434]
Identity'de kaydı olan: 0


In [14]:
before = len(df_trans)
df_trans = df_trans[~dup_mask].reset_index(drop=True)
print(f"{before:,} -> {len(df_trans):,} satır")   

590,540 -> 590,537 satır


In [15]:
# Silme sonrası tabloda tekrar hesaplıyoruz (satır sayısı değişti)
# keep=False: tekrar eden grubun TÜM üyelerini işaretler
key_cols = ["TransactionDT", "TransactionAmt", "card1", "card2", "card3",
            "card4", "card5", "card6", "ProductCD"]
dup_key = df_trans.duplicated(subset=key_cols, keep=False)
print("Anahtar sütunlarda tekrar eden satır sayısı:", dup_key.sum())
print(f"Oran: {dup_key.mean():.3%}")

Anahtar sütunlarda tekrar eden satır sayısı: 102
Oran: 0.017%


In [16]:
# Tekrar eden satırları gruplayıp her grubun büyüklüğüne bakıyoruz
# Çoğu grup 2'li olmalı; 5-10'luk büyük gruplar varsa bunlar ayrıca dikkat çekici
dups = df_trans[dup_key]
group_sizes = dups.groupby(key_cols, dropna=False).size()
print("Grup sayısı:", len(group_sizes))
print(group_sizes.value_counts().sort_index())   # kaç tane 2'li, 3'lü... grup var

# Bu satırlarda isFraud dağılımı: bilgi amaçlı, analiz değil
# Genel fraud oranıyla (~%3,5) kıyaslayıp EDA için not alacağız
print("\nTekrar eden satırlarda isFraud:")
print(dups["isFraud"].value_counts())

Grup sayısı: 41
2    31
3     5
4     3
5     1
8     1
Name: count, dtype: int64

Tekrar eden satırlarda isFraud:
isFraud
0    85
1    17
Name: count, dtype: int64


In [17]:
dup_id = df_id.duplicated(subset=id_cols_wo_id, keep=False)
filled = df_id[id_cols_wo_id].notna().sum(axis=1)

print("Tekrar eden satırlarda ortalama dolu sütun:  ", round(filled[dup_id].mean(), 1))
print("Tekrar etmeyen satırlarda ortalama dolu sütun:", round(filled[~dup_id].mean(), 1))

top_dups = (df_id[dup_id]
            .groupby(["DeviceType", "DeviceInfo", "id_31"], dropna=False)
            .size()
            .sort_values(ascending=False)
            .head(10))
print(top_dups)

Tekrar eden satırlarda ortalama dolu sütun:   8.3
Tekrar etmeyen satırlarda ortalama dolu sütun: 25.9
DeviceType  DeviceInfo  id_31             
NaN         NaN         NaN                   3354
desktop     Windows     edge 16.0              559
                        NaN                     86
            NaN         NaN                     59
            Windows     chrome 63.0             17
                        edge 15.0               14
mobile      NaN         chrome 63.0             12
            iOS Device  mobile safari 11.0      10
desktop     MacOS       safari generic           6
mobile      iOS Device  mobile safari 10.0       6
dtype: int64


**Adım 4: Boşluk analizi**

In [18]:
# Her sütunun boşluk oranını (%) hesaplayan yardımcı fonksiyon
# isna().mean(): her sütunda NaN olan satırların oranı (True=1 sayıldığı için ortalaması oranı verir)
def missing_table(df):
    miss = df.isna().mean() * 100
    # Sadece boşluğu olanları, büyükten küçüğe sıralı göster
    return miss[miss > 0].sort_values(ascending=False).round(2)

miss_trans = missing_table(df_trans)
miss_id = missing_table(df_id)

print(f"Transaction: {len(miss_trans)} / {df_trans.shape[1]} sütunda boşluk var")
print(f"Identity:    {len(miss_id)} / {df_id.shape[1]} sütunda boşluk var")

Transaction: 374 / 394 sütunda boşluk var
Identity:    38 / 41 sütunda boşluk var


In [19]:
# Boşluk oranlarını aralıklara ayırıp kaç sütunun hangi aralıkta olduğuna bakalım
# pd.cut: sürekli değerleri (oranları) önceden belirlediğimiz kovalara dağıtır
bins = [0, 5, 25, 50, 75, 100]
labels = ["0-5%", "5-25%", "25-50%", "50-75%", "75-100%"]

# Boşluğu hiç olmayan sütunlar da dahil olsun diye tüm sütunların oranını alıyoruz
all_trans = df_trans.isna().mean() * 100
all_id = df_id.isna().mean() * 100

# include_lowest=True: 0% olan sütunlar da ilk kovaya girsin
dist = pd.DataFrame({
    "Transaction": pd.cut(all_trans, bins=bins, labels=labels, include_lowest=True).value_counts(),
    "Identity":    pd.cut(all_id, bins=bins, labels=labels, include_lowest=True).value_counts(),
}).reindex(labels)
dist

,Transaction,Identity
0-5%,112,17
5-25%,70,5
25-50%,38,7
50-75%,6,3
75-100%,168,9


In [20]:
# Önceki adımda oluşturduğumuz gruplara göre ortalama boşluk oranı
# Böylece "V sütunları ne kadar boş?" sorusuna tek satırda cevap veriyoruz
# groups sözlüğü Adım 2'den geliyor; bellekte yoksa o hücreyi tekrar çalıştır
for name, cols in groups.items():
    print(f"{name:25s}: ortalama boşluk %{all_trans[cols].mean():.1f}")

# "Diğer" sütunlar için tek tek bakalım, çünkü bunlar analizin ana malzemesi
print("\nDiğer sütunlar:")
print(all_trans[others].round(2).sort_values(ascending=False))

C (sayaç)                : ortalama boşluk %0.0
D (zaman farkı)          : ortalama boşluk %58.2
M (eşleşme T/F)          : ortalama boşluk %49.9
V (Vesta özellikleri)    : ortalama boşluk %43.0
card                     : ortalama boşluk %0.5

Diğer sütunlar:
dist2             93.63
R_emaildomain     76.75
dist1             59.65
P_emaildomain     15.99
addr1             11.13
addr2             11.13
TransactionID      0.00
TransactionDT      0.00
isFraud            0.00
TransactionAmt     0.00
ProductCD          0.00
dtype: float64


In [21]:
# %75'ten fazla boş olan sütunların hangi gruplara dağıldığına bakalım
# all_trans: önceki bloktaki her sütunun boşluk oranı (%)
very_empty = all_trans[all_trans > 75].index

# Her grup için: kaç sütunun %75'ten fazla boş olduğu ve grubun toplam sütun sayısı
# groups sözlüğü Adım 2'den geliyor
for name, cols in groups.items():
    n_empty = len(set(cols) & set(very_empty))
    print(f"{name:25s}: {n_empty:3d} / {len(cols):3d} sütun >%75 boş")

# Gruplara girmeyen sütunlardan (others) kaçı çok boş?
print("\nDiğer sütunlardan çok boş olanlar:",
      [c for c in others if c in very_empty])

C (sayaç)                :   0 /  14 sütun >%75 boş
D (zaman farkı)          :   7 /  15 sütun >%75 boş
M (eşleşme T/F)          :   0 /   9 sütun >%75 boş
V (Vesta özellikleri)    : 159 / 339 sütun >%75 boş
card                     :   0 /   6 sütun >%75 boş

Diğer sütunlardan çok boş olanlar: ['dist2', 'R_emaildomain']


In [22]:
# addr1 ve addr2 aynı satırlarda mı boş?
# (a & b).sum(): ikisi birden boş olan satır sayısı; (a | b).sum(): en az biri boş olan satır sayısı
# Eşitse, ikisi tam olarak aynı satırlarda boş demektir
a1, a2 = df_trans["addr1"].isna(), df_trans["addr2"].isna()
print("İkisi birden boş:", (a1 & a2).sum(), "| en az biri boş:", (a1 | a2).sum())

# Boşluksuz sütunlar: tahminimizi doğrulayalım
complete_trans = df_trans.columns[df_trans.isna().sum() == 0].tolist()
complete_id = df_id.columns[df_id.isna().sum() == 0].tolist()
print("\nTransaction'da boşluksuz:", len(complete_trans), complete_trans)
print("Identity'de boşluksuz:   ", complete_id)

İkisi birden boş: 65706 | en az biri boş: 65706

Transaction'da boşluksuz: 20 ['TransactionID', 'isFraud', 'TransactionDT', 'TransactionAmt', 'ProductCD', 'card1', 'C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9', 'C10', 'C11', 'C12', 'C13', 'C14']
Identity'de boşluksuz:    ['TransactionID', 'id_01', 'id_12']


**Adım 5: Bellek optimizasyonu**

In [24]:
# DataFrame'in bellek kullanımını MB olarak döndüren yardımcı fonksiyon
# deep=True: string sütunlarının gerçek boyutunu da hesaplar
def mem_mb(df):
    return df.memory_usage(deep=True).sum() / 1024**2

In [25]:
# Dönüşümden önce bir karşılaştırma örneği saklıyoruz
# TransactionAmt'ı float32'ye çevirince değer bozuluyor mu, bunu sonra kontrol edeceğiz
orig_amt = df_trans["TransactionAmt"].copy()

def reduce_mem(df):
    """Her sütunu tipine göre en küçük güvenli tipe çevirir."""
    for col in df.columns:
        dtype = df[col].dtype

        if pd.api.types.is_float_dtype(dtype):
            # float64 -> float32: hücre başına 8 byte yerine 4 byte
            # float16'ya inmiyoruz: V sütunlarında büyük değerler var ve hassasiyet kaybı riskli
            df[col] = df[col].astype("float32")

        elif pd.api.types.is_integer_dtype(dtype):
            # downcast="integer": değer aralığına sığan en küçük int tipini seçer
            # (isFraud -> int8, card1 -> int16, TransactionID -> int32 gibi)
            # Bu dönüşüm değer kaybettirmez, çünkü pandas aralığı kontrol eder
            df[col] = pd.to_numeric(df[col], downcast="integer")

        else:
            # Kalan her şey metin sütunu (14 + 17 tane)
            # category: her benzersiz değer bir kez saklanır, satırlar küçük kod tutar
            df[col] = df[col].astype("category")
    return df

before_t, before_i = mem_mb(df_trans), mem_mb(df_id)
df_trans = reduce_mem(df_trans)
df_id = reduce_mem(df_id)
after_t, after_i = mem_mb(df_trans), mem_mb(df_id)

print(f"Transaction: {before_t:,.1f} MB -> {after_t:,.1f} MB ({1 - after_t/before_t:.1%} azalma)")
print(f"Identity:    {before_i:,.1f} MB -> {after_i:,.1f} MB ({1 - after_i/before_i:.1%} azalma)")

Transaction: 1,791.7 MB -> 861.1 MB (51.9% azalma)
Identity:    56.5 MB -> 16.0 MB (71.7% azalma)


In [26]:
# Doğrulama 1: TransactionAmt'ta float32 dönüşümü ne kadar sapma yarattı?
# Sapma kuruşun çok altındaysa sorun yok
diff = (df_trans["TransactionAmt"].astype("float64") - orig_amt).abs().max()
print("TransactionAmt en büyük sapma:", diff)

# Doğrulama 2: yeni tipler
print(df_trans.dtypes.value_counts())
print(df_trans[["TransactionID", "isFraud", "TransactionDT", "card1"]].dtypes)

TransactionAmt en büyük sapma: 0.0003749999996216502
float32     376
category      8
int32         2
int8          1
category      1
int16         1
category      1
category      1
category      1
category      1
category      1
Name: count, dtype: int64
TransactionID    int32
isFraud           int8
TransactionDT    int32
card1            int16
dtype: object


**Adım 6: Birleştirme ve kaydetme**

In [27]:
# Anahtar sütunun tipi iki tabloda aynı mı? (ikisi de int32 olmalı)
print(df_trans["TransactionID"].dtype, df_id["TransactionID"].dtype)

# left join: transaction tablosunu ana tablo alıp identity'yi üzerine ekliyoruz
# validate="one_to_one": ID'ler her iki tabloda benzersiz olmalı, değilse pandas hata verir
# (Adım 2'de unique olduklarını doğrulamıştık, bu ikinci bir güvence)
df = df_trans.merge(df_id, on="TransactionID", how="left", validate="one_to_one")

# Doğrulamalar: assert koşul False ise hata fırlatır, True ise sessiz geçer
assert len(df) == 590_537, "Satır sayısı değişti!"        # left join satır eklememeli/silmemeli
assert df.shape[1] == 394 + 41 - 1, "Sütun sayısı beklenenden farklı!"   # TransactionID ortak, bir kez sayılır
print("Birleştirme sonrası shape:", df.shape)

int32 int32
Birleştirme sonrası shape: (590537, 434)


In [28]:
# Kimlik kaydı olan işlem oranı hâlâ %24,4 mü? (id_01 sadece identity'den geliyor)
# id_01 dolu olan satırlar = identity ile eşleşen işlemler
print(f"Identity ile eşleşen işlem oranı: {df['id_01'].notna().mean():.1%}")

# Birleştirme sonrası bellek ve tip kontrolü
print(f"Birleşik tablo: {mem_mb(df):,.1f} MB")
print(df.dtypes.astype(str).value_counts())

Identity ile eşleşen işlem oranı: 24.4%
Birleşik tablo: 924.2 MB
float32     399
category     31
int32         2
int8          1
int16         1
Name: count, dtype: int64


In [29]:
# data/ klasörüne kaydediyoruz (.gitignore'da olduğundan emin ol, dosya yüzlerce MB olabilir)
OUT_PATH = "../data/train_merged.parquet"
df.to_parquet(OUT_PATH, index=False)

# Doğrulama: dosyayı geri okuyup tiplerin ve boyutun korunduğunu kontrol et
df_check = pd.read_parquet(OUT_PATH)
print("Okunan shape:", df_check.shape)
print("Tipler aynı mı?:", (df.dtypes == df_check.dtypes).all())
print(f"Bellekte: {mem_mb(df_check):,.1f} MB")

import os
print(f"Dosya boyutu: {os.path.getsize(OUT_PATH) / 1024**2:,.1f} MB")

Okunan shape: (590537, 434)
Tipler aynı mı?: True
Bellekte: 924.2 MB
Dosya boyutu: 80.2 MB


In [30]:
del df_trans, df_id

# 01 – Data Overview: Özet ve Bulgular

Bu notebook'un amacı, analize başlamadan önce veriyi **tanımak, temizlik açısından kontrol etmek, küçültmek ve tek bir dosyada toplamak**.

---

## 1. Veri seti nedir?

**IEEE-CIS Fraud Detection** (Kaggle yarışması, veriyi sağlayan şirket: Vesta). Online kart işlemlerinden oluşuyor ve her işlem için "dolandırıcılık mı (`isFraud = 1`), değil mi (`isFraud = 0`)" bilgisi var.

Veri iki ayrı dosyada geliyor:

| Dosya | İçerik | Satır × Sütun | Bellek |
|---|---|---|---|
| `train_transaction.csv` | İşlemin kendisi: tutar, kart, adres, e-posta, ürün tipi, anonim özellikler | 590.540 × 394 | 1.791,7 MB |
| `train_identity.csv` | İşlemi yapan cihaz/tarayıcı bilgisi | 144.233 × 41 | 56,5 MB |

İki dosya `TransactionID` sütunuyla birbirine bağlanıyor.

## 2. Sütunlar neyi anlatıyor?

**Transaction tablosu (394 sütun):**

| Grup | Sütun sayısı | Anlamı |
|---|---|---|
| `TransactionID` | 1 | İşlemin benzersiz numarası |
| `isFraud` | 1 | Hedef değişken (0 = normal, 1 = fraud) |
| `TransactionDT` | 1 | Belirsiz bir başlangıç anından itibaren geçen süre (saniye). **Gerçek bir tarih değil.** |
| `TransactionAmt` | 1 | İşlem tutarı |
| `ProductCD` | 1 | Ürün tipi (5 farklı değer) |
| `card1`–`card6` | 6 | Kart bilgileri (kart tipi, banka vb.) |
| `addr1`, `addr2` | 2 | Adres bilgileri |
| `dist1`, `dist2` | 2 | Mesafe bilgileri |
| `P_emaildomain`, `R_emaildomain` | 2 | Alıcı / gönderici e-posta alan adları |
| `C1`–`C14` | 14 | Sayaç türü özellikler (anlamları gizli) |
| `D1`–`D15` | 15 | Zaman farkı türü özellikler (anlamları gizli) |
| `M1`–`M9` | 9 | Eşleşme (T/F) bayrakları |
| `V1`–`V339` | 339 | Vesta'nın kendi ürettiği özellikler (anlamları gizli) |

**Identity tablosu (41 sütun):** `TransactionID`, `id_01`–`id_38` (anonim cihaz/ağ/tarayıcı özellikleri), `DeviceType` (mobil/masaüstü), `DeviceInfo` (cihaz modeli).

> **Sınırlılık:** Yarışma verisi gizlilik için anonimleştirilmiş. C, D, M, V ve `id_*` sütunlarının gerçek anlamı bilinmiyor. Gerçek bir fintech'te bu sütunların ne olduğunu bilirdik; burada bunlar "anlamı bilinmeyen ham sinyal" olarak ele alınıyor. Verinin %86'sı (339/394) V sütunlarından oluşuyor.

**Veri tipleri (transaction, ham hali):** 376 sütun `float64`, 14 sütun metin, 4 sütun `int64`. Bazı tam sayı olması gereken sütunlar (`card2`, `addr1` gibi) boşluk içerdikleri için ondalıklı okundu.

## 3. İki tablonun ilişkisi

`TransactionID` üzerinden yapılan kontroller:

- Her iki tabloda da `TransactionID` **benzersiz** (bir işlem tek satır).
- Identity'deki ID'lerin **%100'ü** transaction tablosunda var (kimliği olup işlemi olmayan kayıt yok).
- İşlemlerin yalnızca **%24,4'ünün** kimlik (cihaz) kaydı var, yaklaşık dörtte üçünün yok.

**Birleştirme:** Birleştirmede `left join` (transaction ana tablo). `inner join` işlemlerin %75,6'sını silerdi. Ayrıca "kimlik kaydı olmaması" başlı başına bir bilgi olabilir; EDA'da fraud oranıyla ilişkisine bakılacak.

## 4. Tekrarlı kayıt analizi

**a) Transaction'da birebir aynı satırlar**
- `TransactionID` hariç 393 sütunun hepsi aynı olan **3 çift** bulundu.
- Üç çiftte de ID'ler ardışık (örn. 3282456/3282457) ve `TransactionDT` aynı (aynı saniye); üçü de `isFraud = 0`.
- Yorum: Sistemin aynı işlemi iki kez kaydetmesi gibi görünüyor. Bu bir yorum; kesin kanıt değil.
- Her çiftin ikinci kopyası silindi (3 satır). Silinen ID'lerin identity kaydı olmadığı için kimlik bilgisi kaybı yok. Satır sayısı **590.540 → 590.537**.

**b) Identity'de "tekrar eden" satırlar**
- ID hariç diğer sütunları aynı olan **3.767 satır** var.
- Tekrar eden satırlarda ortalama **8,3 sütun** dolu; tekrar etmeyenlerde **25,9**. Yani tekrarlar çoğunlukla bilgisi az (çok boş) kayıtlar. Geri kalanı yaygın cihaz/tarayıcı kombinasyonları (aynı cihazdan birden fazla işlem).
- Her satır farklı bir işleme ait; bunlar gerçek kopya değil. Bu nedenle silinmedi.

**c) "İş mantığı" tekrarı (aynı zaman + tutar + kart + ürün)**
- **102 satır** (%0,017), **41 grup** (31 ikili, 5 üçlü, 3 dörtlü, 1 beşli, 1 sekizli).
- Bu satırlar diğer sütunlarda farklı, yani birebir kopya değil.
- Bu 102 satırın **17'si fraud (%16,7)**.
- EDA'da incelenecek ve feature engineering için aday bir sinyal ("aynı kartın kısa sürede aynı tutarı tekrar denemesi").

## 5. Boşluk (missing value) analizi

**Genel durum**
- Transaction: 394 sütunun **374'ünde** boşluk var. Boşluksuz 20 sütun: `TransactionID`, `isFraud`, `TransactionDT`, `TransactionAmt`, `ProductCD`, `card1` ve `C1`–`C14`.
- Identity: 41 sütunun **38'inde** boşluk var. Boşluksuz: `TransactionID`, `id_01`, `id_12`.

**Boşluk oranı dağılımı (sütun sayısı)**

| Boşluk oranı | Transaction | Identity |
|---|---|---|
| %0–5 | 112 | 17 |
| %5–25 | 70 | 5 |
| %25–50 | 38 | 7 |
| %50–75 | 6 | 3 |
| %75–100 | **168** | 9 |

Sütunlar ya çok dolu ya çok boş; orta bölge neredeyse boş. Bu, boşluğun rastgele olmadığını düşündürüyor (bazı sütunlar yalnızca belirli koşullarda doluyor olabilir).

**Gruplara göre ortalama boşluk (transaction)**

| Grup | Ortalama boşluk |
|---|---|
| C | %0,0 |
| card | %0,5 |
| V | %43,0 |
| M | %49,9 |
| D | %58,2 |

**Anlamı bilinen sütunlar**

| Sütun | Boşluk |
|---|---|
| `dist2` | %93,63 |
| `R_emaildomain` | %76,75 |
| `dist1` | %59,65 |
| `P_emaildomain` | %15,99 |
| `addr1`, `addr2` | %11,13 (ikisi de) |

**Önemli gözlemler**
- %75'ten fazla boş olan 168 sütunun dağılımı: **159 V, 7 D, `dist2`, `R_emaildomain`**. V grubunun ortalaması %43 olsa da neredeyse yarısı çok boş; V bir blok değil, farklı doluluktaki kümelerden oluşuyor.
- `addr1` ve `addr2` **tam olarak aynı 65.706 satırda** birlikte boş. Adres bilgisi muhtemelen bir bütün olarak ya giriliyor ya girilmiyor.
- Hiçbir sütun silinmedi. Fraud verisinde "bu alan neden boş?" bilgi taşıyabilir; EDA'da boşluk örüntüleri `isFraud` ile ilişkilendirilecek. Bellek açısından da silmeye gerek kalmadı.

## 6. Bellek optimizasyonu

Birleştirmeden önce her tablo ayrı ayrı küçültüldü:

| Dönüşüm | Neden |
|---|---|
| `float64` → `float32` (376 + 23 sütun) | Hücre başına 8 yerine 4 byte. `float16`'ya inilmedi: hassasiyet kaybı riski var. |
| Tam sayılar → en küçük uygun `int` | `isFraud` → `int8`, `card1` → `int16`, `TransactionID`/`TransactionDT` → `int32`. Değer kaybı yok. |
| Metin → `category` (14 + 17 sütun) | Kardinalite düşük (en yüksek: `DeviceInfo` 1.786 farklı değer). Her benzersiz değer bir kez saklanır. |

**Sonuç**

| Tablo | Önce | Sonra | Azalma |
|---|---|---|---|
| Transaction | 1.791,7 MB | 861,1 MB | %51,9 |
| Identity | 56,5 MB | 16,0 MB | %71,7 |

**Doğrulama:** `TransactionAmt` için `float32` dönüşümünün yarattığı en büyük sapma **0,000375**, yani tutar analizi için ihmal edilebilir. Kalan boyutun neredeyse tamamı 376 `float32` sütunundan geliyor.

## 7. Birleştirme ve kaydetme

- `left join`, `TransactionID` üzerinden, `validate="one_to_one"` ile (ID'ler benzersiz olmalı, değilse hata verir).
- Doğrulamalar: satır sayısı **590.537** (değişmedi), sütun sayısı **434** (394 + 41 − 1 ortak ID). Identity ile eşleşen işlem oranı hâlâ **%24,4**.
- Birleşik tablo bellekte **924,2 MB** (identity 144 bin satırdan 590 bin satıra yayıldığı için iki küçük tablonun toplamından biraz yüksek).
- Kayıt formatı **Parquet** (`data/train_merged.parquet`, **80,2 MB**). CSV yerine Parquet seçildi çünkü CSV tip bilgisini saklamaz (`float32` ve `category` tekrar okununca eski haline döner, optimizasyon boşa gider). Parquet tipleri korur, daha küçük ve çok daha hızlı okunur. Dosya geri okunup tiplerin aynı olduğu doğrulandı.

## 8. Alınan kararların özeti

| Konu | Karar | Gerekçe |
|---|---|---|
| Birleştirme türü | `left join` | İşlemlerin %75,6'sını kaybetmemek; "kimlik yok" bilgisini korumak |
| 3 birebir aynı transaction satırı | Silindi | Ardışık ID, aynı saniye, aynı değerler: sistem kaynaklı çift kayıt varsayımı |
| Identity'deki 3.767 "tekrar" | Silinmedi | Boş kayıtlar ve paylaşılan cihazlar, gerçek kopya değil |
| 102 satırlık iş mantığı tekrarı | Silinmedi | Birebir kopya değil, olası fraud sinyali |
| Boş sütunlar | Silinmedi | Boşluğun kendisi bilgi taşıyabilir; EDA'da test edilecek |
| Tip dönüşümü | float32 / küçük int / category | Bellek yarıya indi, değer kaybı ihmal edilebilir |
| Kayıt formatı | Parquet | Tipleri korur, küçük ve hızlı |

## 9. EDA notebook'una taşınan sorular

1. Genel fraud oranı nedir, sınıf dengesizliği ne kadar? (~%3,5 beklentisi doğrulanacak)
2. Kimlik kaydı olan ve olmayan işlemlerde fraud oranı farklı mı? (`has_identity`)
3. Adres (`addr1`/`addr2`), `dist` ve `R_emaildomain` boşlukları fraud ile ilişkili mi? Boşluk belirli bir `ProductCD`'ye mi bağlı?
4. Tekrar eden işlemlerde (102 satır) fraud oranı gerçekten yüksek mi?
5. `TransactionAmt` dağılımı fraud ve normal işlemlerde farklı mı?
6. `TransactionDT`'den türetilen günün saati / haftanın günü örüntüleri var mı? (Gerçek tarih değil, ama göreli zaman örüntüleri çıkarılabilir.)
7. `ProductCD`, `card4`, `card6` ve e-posta alan adı kırılımlarında fraud oranları nasıl değişiyor?

## 10. Bu notebook'un sınırlılıkları

- C, D, M, V ve `id_*` sütunlarının anlamı gizli; yorumlar bu sütunlar için sınırlı kalıyor.
- `TransactionDT` gerçek bir tarih değil, sadece göreli zaman.
- Tekrarlı kayıtlara dair yorumlar (sistem kaynaklı kopya, kart test denemesi) veriden çıkarılmış varsayımlar; kesin bilgi değil.
- Boşlukların *nedeni* bilinmiyor; EDA'da fraud ve ürün tipiyle ilişkilendirilerek test edilecek.